src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1641,7 +1643,14 @@ TargetShardIntervalForModify(Query *query)
 	}
 
 	prunedShardCount = list_length(prunedShardList);
-	if (prunedShardCount != 1)
+	if (updateOrDelete && prunedShardCount > 1 && prunedShardCount == shardCount)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("distributed modifications must target exactly one "
+							   "shard"),
+						errhint("Consider using equality filter on partition column.")));
```

Comment:
Might be nice to even elaborate a bit further here, e.g. look up the partition column name and use `"Consider using equality filter on partition column (%s = ...).", partitionColumnName`